common/beerocks/bcl/include/bcl/network/socket_refactorings.h

```diff
@@ -0,0 +1,795 @@
+    {
+    }
+};
+
+class UdsServerSocket : public ServerSocketAbstractImpl {
+public:
+    explicit UdsServerSocket(const std::shared_ptr<UdsSocket> &socket)
+        : ServerSocketAbstractImpl(socket)
+    {
+    }
+};
+
+class UdsClientSocket : public ClientSocketAbstractImpl {
+public:
+    explicit UdsClientSocket(const std::shared_ptr<UdsSocket> &socket)
+        : ClientSocketAbstractImpl(socket)
+    {
+    }
+};
+
+/**
+ * prplmesh-democracy: Do you prefer templates?
+ * These template classes would replace previous ServerSocket and ClientSocket derived classes
+ */
+template <class SocketType> class ServerSocketImpl : public ServerSocketAbstractImpl {
+public:
+    explicit ServerSocketImpl(const std::shared_ptr<SocketType> &socket)
+        : ServerSocketAbstractImpl(socket)
+    {
+    }
+};
+
+template <class SocketType> class ClientSocketImpl : public ClientSocketAbstractImpl {
+public:
+    explicit ClientSocketImpl(const std::shared_ptr<SocketType> &socket)
+        : ClientSocketAbstractImpl(socket)
+    {
+    }
+};
+
+} // namespace sockets
+
+/**
+ * A CMDU handler class is the equivalent in functionality to, for example, each of those methods
+ * found in the backhaul manager that are used to handle the different types of messages.
+ *
+ * To test a particular implementation of CMDU handler, we must know which are its dependencies and
+ * what does the handler do with them. All dependencies must be injected as parameters of the
+ * handler constructor and the handler code must be programmed to interfaces, not to particular
+ * implementations. A dependency that most if not all CMDU handlers will have is the socket
+ * connection that the message was received on and that will be used to send the CMDU replies.
+ * This way we can write tests for each implementation of the CMDU handler just by using mocks of
+ * the dependencies and setting expectations on the mocks, that is, without sending any single
+ * byte on a real socket connection.
+ */
+class CmduHandler {
+public:
+    virtual ~CmduHandler()                                          = default;
+    virtual void handle_message(ieee1905_1::CmduMessageRx &cmdu_rx) = 0;
+};
+
+/**
+ * A message dispatcher is also a message handler so it can be used "instead of" a message handler.
+ * This class is a replacement for which otherwise would be a very long switch-case statement. See
+ * an example of such a switch statement in backhaul_manager::handle_1905_1_message()
+ */
+class CmduDispatcher : public CmduHandler {
+public:
+    /**
+     * @brief Registers a handler for messages of given type.
+     *
+     * @param message_type The type of CMDU.
+     * @param handler Message handler to register.
+     * @return True on success and false otherwise (i.e.: there is an already registered handler
+     * for messages of given type)
+     */
+    bool register_handler(ieee1905_1::eMessageType message_type,
+                          const std::shared_ptr<CmduHandler> &handler)

```

Comment:
I think it would be more appropriate here to pass the handler as a function rather than an object. E.g. you could pass the handle_foo members of backhaul manager directly, rather than requiring a new class for each handler.

Funny, for EventLoop we were both taking exactly the opposite side... Which makes me think, maybe I'm wrong, because we can make a backhaul manager internal class that inherits from CmduHandler and that takes a function as a constructor argument.

Reply:
Right, this is something yet to be agreed-upon.  I'll write two examples, one with each approach and a list of pros and cons, so we can decide.